modules/indexing/src/main/java/org/apache/ignite/internal/visor/cache/index/IndexRebuildTask.java

```diff
@@ -0,0 +1,190 @@
+
+    /** {@inheritDoc} */
+    @Override protected IndexRebuildJob job(IndexRebuildTaskArg arg) {
+        return new IndexRebuildJob(arg, debug);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected @Nullable IndexRebuildTaskRes reduce0(List<ComputeJobResult> results) throws IgniteException {
+        Map<UUID, IndexRebuildJobRes> taskResultMap = results.stream()
+            .collect(Collectors.toMap(res -> res.getNode().id(), ComputeJobResult::getData));
+
+        return new IndexRebuildTaskRes(taskResultMap);
+    }
+
+    /** */
+    private static class IndexRebuildJob extends VisorJob<IndexRebuildTaskArg, IndexRebuildJobRes> {
+        /** */
+        private static final long serialVersionUID = 0L;
+
+        /**
+         * Create job with specified argument.
+         *
+         * @param arg Job argument.
+         * @param debug Flag indicating whether debug information should be printed into node log.
+         */
+        protected IndexRebuildJob(@Nullable IndexRebuildTaskArg arg, boolean debug) {
+            super(arg, debug);
+        }
+
+        /** {@inheritDoc} */
+        @Override protected IndexRebuildJobRes run(@Nullable IndexRebuildTaskArg arg) throws IgniteException {
+            assert arg.cacheToIndexes() != null && !arg.cacheToIndexes().isEmpty() : "Cache to indexes map must be specified.";
+
+            Set<String> notFound = new HashSet<>();
+
+            GridCacheProcessor cacheProcessor = ignite.context().cache();
+
+            Map<String, Set<String>> cacheToIndexes = new HashMap<>();
+            Map<String, Set<String>> cacheToMissedIndexes = new HashMap<>();
+
+            for (Entry<String, Set<String>> indexesByCache : arg.cacheToIndexes().entrySet()) {
+                String cache = indexesByCache.getKey();
+                Set<String> indexesArg = indexesByCache.getValue();
+                int cacheId = CU.cacheId(cache);
+
+                GridCacheContext<?, ?> cacheCtx = cacheProcessor.context().cacheContext(cacheId);
+
+                if (cacheCtx == null) {
+                    notFound.add(cache);
+                    continue;
+                }
+
+                Set<String> existingIndexes = indexes(cache);
+                Set<String> indexesToRebuild = cacheToIndexes.computeIfAbsent(cache, s -> new HashSet<>());
+                Set<String> missedIndexes = cacheToMissedIndexes.computeIfAbsent(cache, s -> new HashSet<>());
+
+                if (indexesArg.isEmpty())
+                    indexesToRebuild.addAll(existingIndexes);
+                else {
+                    indexesArg.forEach(index -> {
+                        if (!existingIndexes.contains(index)) {
+                            missedIndexes.add(index);
+                            return;
+                        }
+
+                        indexesToRebuild.add(index);
+                    });
+                }
+            }
+
+            if (hasIndexes(cacheToIndexes)) {
```

Comment:
Why not **!cacheToIndexes.isEmpty()**?

Reply:
Because it checks collections that are values of the map, not the map itself